### DISCUSSION TITLE
Using DICOM folders of data instead of nifti files for image volumes.

### DISCUSSION THREAD

**Discussion by stevenagl12** [Project-MONAI/MONAILabel#1008] [Q&A]
Hi, I have a quick question. So if you have a directory architecture where you have a folder for each patient that contains all of the image data for that patient, can you use a dicom series reader such as SimpleITK to parse these dicom files into an image volume for building a dataset for monai label? Is there currently a script for this or would it need to be specialized in the config or inference scripts for monailabel.

Also, if you have a separate folder of unannotated images that you would like to run inference on, how do you specific this?

**Comment by lassoan** [comment]
There are often several ways to interpret a set of DICOM files in a folder and select what series are relevant.

If you are feeling lucky and only interested in image (not segmentations, structure sets, etc.) then you can use tools such as dcm2niix to extract most likely interesting nrrd or nifti images.

If you have more heterogeneous data set (that has not been filtered for removing duplicate, low-quality, potentially incorrectly named, oriented, etc.) then I would recommend to iterate through all the studies and visually verify data quality and select relevant series. You can do this in 3D Slicer by drag-and-dropping the entire folder structure into 3D Slicer, then loading them using the DICOM module and export the relevant data objects to research file formats. 3D Slicer can import not just 3D images but also 4D image sequences and DICOM Segmentation Objects, RT Structure Sets, and many other data types that you might need.

**Reply by stevenagl12** [reply]
I have already verified these images are correct, using a DICOM web browser, but I don't want to have to write a script to now convert each instance into a nifti file format. When I downladed them, they have a set format where I have the patient name, then an IMAGES subfolder with a DICOMDIR file, and then the series files under the IMAGES folder. I also have annotations for many of them and am wondering if I could create some sort of parser like a dataloader that works with monai label to pull in the annotation file?

**Reply by lassoan** [reply]
You can export the entire Slicer DICOM database content to nrrd (or nifti, etc.) file format with a script like this. It can be much simpler, but I wanted to add an example of how to filter what you want to save and construct the filename from DICOM tags:

```python
from DICOMLib import DICOMUtils
outputFolder = "c:/tmp/exptest/"
patientUIDs = slicer.dicomDatabase.patients()
for patientUID in patientUIDs:
    loadedNodeIDs = DICOMUtils.loadPatientByUID(patientUID)
    for loadedNodeID in loadedNodeIDs:
        # Check if we want to save this node
        node = slicer.mrmlScene.GetNodeByID(loadedNodeID)
        if not node or not node.IsA('vtkMRMLScalarVolumeNode'):
            continue
        # Construct filename
        shNode = slicer.mrmlScene.GetSubjectHierarchyNode()
        seriesItem = shNode.GetItemByDataNode(node)
        studyItem = shNode.GetItemParent(seriesItem)
        patientItem = shNode.GetItemParent(studyItem)
        filename = shNode.GetItemAttribute(patientItem, 'DICOM.PatientID')
        filename += '_' + shNode.GetItemAttribute(studyItem, 'DICOM.StudyDate')
        filename += '_' + shNode.GetItemAttribute(seriesItem, 'DICOM.SeriesNumber')
        filename += '_' + shNode.GetItemAttribute(seriesItem, 'DICOM.Modality')
        filename = slicer.app.ioManager().forceFileNameValidCharacters(filename) + ".nrrd"
        # Save node
        print(f'Write {node.GetName()} to {filename}')
        success = slicer.util.saveNode(node, outputFolder+filename)
    slicer.mrmlScene.Clear()
```

**Reply by stevenagl12** [reply]
I just tried your code, and got the following error: AttributeError: qSlicerIOManager has no attribute named 'forceFileNameValidCharacters'

**Reply by lassoan** [reply]
You need to use a recent Slicer version (Slicer-5.x). `forceFileNameValidCharacters` may have not been exposed in earlier versions. You can also remove characters that are not permitted in filenames with some Python string replace.

**Reply by stevenagl12** [reply]
This helped a lot thanks.

**Comment by tangy5** [comment]
Hi @stevenagl12 , the current sample apps starts and creates workflow based on data load such as https://github.com/Project-MONAI/MONAILabel/blob/5a7cbd8d5de945a99085d0f6495de21485c47415/sample-apps/radiology/lib/infers/segmentation_spleen.py#L56

You can try specify image reader such as ITKreader, there is a parameter can set DICOM series. 
See **series_name** in https://docs.monai.io/en/stable/data.html#itkreader
You can create a customized transformation in  https://github.com/Project-MONAI/MONAILabel/blob/5a7cbd8d5de945a99085d0f6495de21485c47415/sample-apps/radiology/lib/infers/segmentation_spleen.py#L54
for infer or trainer and load image as preferred. Thanks for using MONAI Label

**Comment by diazandr3s** [comment]
Nice discussion :)

Currently, the radiology app in MONAI Label supports the [local file archive datastore](https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/datastore/local.py) in which images have to be in nrrd, nifti or compressed nifti format OR in a [DICOM Web server](https://github.com/Project-MONAI/MONAILabel/blob/main/monailabel/datastore/dicom.py) (i.e. Orthanc)

This means that if you want the images to remain on a DICOM Web server, I suggest you use MONAI Label like this: 

`monailabel start_server --app apps/radiology --studies IP_TO_DICOM_WEB_SERVER`

To this question:

> Also, if you have a separate folder of unannotated images that you would like to run inference on, how do you specific this?

If the unlabelled images are in DICOM format, my suggestion is to put them in a DICOM Web server and then run batch inference. Otherwise, convert them into NRRD format, start the MONAI Label server pointing to that folder (local datastore) and then run batch inference. i.e. `monailabel start_server --app apps/radiology --studies FOLDER_UNLABELED_IMAGES --conf models deepedit`


Hope this helps,

**Reply by lassoan** [reply]
FYI, [Slicer offers access to its DICOM database via a DICOMweb endpoint](https://slicer.readthedocs.io/en/latest/user_guide/modules/webserver.html#dicomweb-endpoints). So, if you want to fetch data from a remote DICOM server then you don't need to set up a complicated server like Orthanc, but you can simply install Slicer on the remote computer, drag-and-drop DICOM folders to Slicer application window (to import them into Slicer's DICOM database), and then enable `DICOMweb API` in the Slicer's Web Server module.

**Reply by stevenagl12** [reply]
That is awesome, thank you.
